### ISSUE TITLE
TraitsImplementing calls WithInterface not ActorsWithTrait(?)

### ISSUE THREAD

**Issue by anvilvapre** [OpenRA/OpenRA#13555]
possible bug in bleed, 
depending on how it was designed to work,

background:
- TraitContainer.Get and GetOrDefault throws an exception when a trait is queried for an actor and that trait does not exist. 
- TraiContainer.GetMultiple returns an enumeration, will search the complete container but does not check if it at least finds one entry. So here the same exception is not thrown. 
- When I log all traits that are being added to the TraitContainers there is no trait added for the World actor with the name IAutoSelectionSize. tc01, tc02, etc actors seem to have it

problem:
- After changing the TraitContainer with a new implementation the game crashes with the error saying that Actor 0 does not have trait of type `OpenRA.Traits.IAutoSelectionSize'. The exception is triggered in GetMultiple - where I also added the check. 

cause:
- The check in GetMultiple in my implemention. 
- The original GetMultiple implementation that will result in an empty enumeration. 
- World being initialized where DetermineBounds is called. 
- World.DetermineBounds calling World.TraitsImplementing
- World.TraitsImplementing call TraitDictionary.WithInterface with this as actor argument.
- TraitDictionary.WithInterface only returning traits for that specific actor. 

solution:
- Should perhaps World.TraitsImplementing not call ActorsWithTrait instead? Returning all actors with that trait?
- Or, the DetermineBounds incorrectly calls TraitsImplementing 

in summary:
- i think WithInterface is called where ActorsWithTrait is meant? 
- WithInterface is also a misleading name? Perhaps better to call it with Trait?

bug:
- traits being queried that will always result in an empty result 

or is it dependent on the game mod being loaded whether an actor may or may not implement a certaint trait and therefore it is acceptable to query an actor for a trait it does not have?

**Comment by WolfGaming** [comment]
- TraitContainer.Get and GetOrDefault throws an exception when a trait is queried for an actor and that trait does not exist. 
_**This is probably a bug if GetOrDefault throws an exception.**_
- TraiContainer.GetMultiple returns an enumeration, will search the complete container but does not check if it at least finds one entry. So here the same exception is not thrown. 
_**Probably intended.**_
- When I log all traits that are being added to the TraitContainers there is no trait added for the World actor with the name IAutoSelectionSize. tc01, tc02, etc actors seem to have it
_**IAutoSelectionSize is probably an interface based on the name so that might have something to do with why you can't find it, don't take my word for it though been a while since I looked at that code.**_

problem:
- After changing the TraitContainer with a new implementation the game crashes with the error saying that Actor 0 does not have trait of type `OpenRA.Traits.IAutoSelectionSize'. The exception is triggered in GetMultiple - where I also added the check. 

_**As said before, GetMultiple is supposed to return an empty collection if none exist.**_

cause:
- The check in GetMultiple in my implemention. 
- The original GetMultiple implementation that will result in an empty enumeration. 
- World being initialized where DetermineBounds is called. 
- World.DetermineBounds calling World.TraitsImplementing
- World.TraitsImplementing call TraitDictionary.WithInterface with this as actor argument.
- TraitDictionary.WithInterface only returning traits for that specific actor. 

solution:
- Should perhaps World.TraitsImplementing not call ActorsWithTrait instead? Returning all actors with that trait?
- Or, the DetermineBounds incorrectly calls TraitsImplementing 

_**DetermineBounds is working as intended and it would be pretty senseless to expect a world to initialize itself without knowing how big it is.**_
_**Calling TraitsImplementing is always ok anywhere in the code and 'World' is a special kind of actor but an actor nonetheless as it can have traits.**_

in summary:
- i think WithInterface is called where ActorsWithTrait is meant? 
- WithInterface is also a misleading name? Perhaps better to call it with Trait?

_**You are probably misunderstanding how TraitsImplementing is supposed to work, it is supposed to get all traits that implement a specific interface, not a specific class.  Putting a check in GetMultiple makes no sense as returning an empty collection is cleaner (You would have to either do a try catch or do a GetDefault before GetMultiple which just overcomplicates things) and if you really care about it you can either look for a specific one or check the collection yourself.**_

_**The only thing I see wrong with this whole thing is if GetDefault throws an exception as it is not supposed to.**_

**Comment by anvilvapre** [comment]
Thanks for the explanation. 
GetOrDefault throws only an exception if multiple traits are there. Not if there are none. 
I misread. I also thought DetermineBounds was in World, not in Actor.
Please close this issue, I don't seem to be able to do it myself.

**Comment by GraionDilach** [comment]
GetOrDefault throwing on multiple instances is intended IMO.
